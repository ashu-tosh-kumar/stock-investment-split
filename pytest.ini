[pytest]
testpaths = test
addopts = -p no:warnings

markers =
    integration: marks tests as integration (deselect with '-m "not integration"')
