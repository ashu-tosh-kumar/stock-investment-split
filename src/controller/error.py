from typing import Callable

import numpy as np

from src.models.financial_instruments import FinancialInstruments


def mean_squared_error(financial_instruments: FinancialInstruments, new_investment: float) -> Callable[[np.ndarray], float]:
    """Closure that returns a callable function that when passed with `N_i` values i.e.
    the target investment ratio for financial instruments returns the Mean Squared Error
    (`MSE`)

    Equation: `(1/n)sum_(i=1)^n ((I_i + N_i)/(SI + SN) - T_i)^2`
    Use https://asciimath.org to render above equation

    Args:
        financial_instruments (FinancialInstruments): Model representing all financial
        instruments part of the portfolio
        new_investment (float): Total amount of new investment to be made

    Returns:
        Callable[[np.ndarray], float]: Returns a callable function that takes `N_i` and
        returns `MSE`
    """
    num_instruments: int = len(financial_instruments.instruments)
    I_i: np.ndarray = np.asarray([instrument.initial_investment for instrument in financial_instruments.instruments])  # /NOSONAR
    SI: float = sum([instrument.initial_investment for instrument in financial_instruments.instruments])
    SN: float = new_investment
    T_i: np.ndarray = np.asarray([instrument.target_investment_ratio for instrument in financial_instruments.instruments])  # /NOSONAR

    def mean_squared_error_inner(N_i: np.ndarray) -> float:  # /NOSONAR
        """Inner function that actually computes the MSE and returns the same

        Args:
            N_i (np.ndarray): Array of new investment allocated to each financial
            instrument

        Returns:
            float: MSE
        """
        I_i_plus_N_i = np.add(I_i, N_i)  # /NOSONAR
        SI_plus_SN = SI + SN  # /NOSONAR
        I_i_plus_N_i_divided_by_SI_plus_SN = np.divide(I_i_plus_N_i, SI_plus_SN)  # /NOSONAR
        se_array = np.power(np.subtract(I_i_plus_N_i_divided_by_SI_plus_SN, T_i), 2)
        mse: float = np.sum(se_array) / num_instruments
        return mse

    return mean_squared_error_inner
