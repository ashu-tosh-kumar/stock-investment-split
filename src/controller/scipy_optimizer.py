from typing import Callable

import numpy as np
import scipy

from src.models.financial_instruments import FinancialInstruments
from src.models.optimizers import ScipyOptimizeMinimizeMethods


def _constraint_between_0_and_1(x: np.ndarray) -> bool:
    """Constraint: Each element of x must be between 0 and 1

    Args:
        x (np.ndarray): Array of values

    Returns:
        bool: Returns whether the constraint is satisfied or not
    """
    return np.all(np.logical_and(x >= 0, x <= 1))  # type: ignore


def _constraint_sum_equal_to_1(x: np.ndarray) -> float:
    """Constraint: The sum of all elements of x must be equal to 1

    Args:
        x (np.ndarray): Array of values

    Returns:
        float: Returns if constrained is satisfied or not
    """
    return x.sum() - 1


def scipy_optimize_minimize(
    error_function: Callable[[FinancialInstruments, float], Callable[[np.ndarray], float]],
    financial_instruments: FinancialInstruments,
    new_investment: float,
    method: ScipyOptimizeMinimizeMethods = ScipyOptimizeMinimizeMethods.SLSQP,
) -> np.ndarray:
    """Optimizer using `scipy.optimize.minimize` to minimize a the `error_function`

    Args:
        error_function (Callable[[FinancialInstruments, float], Callable[[np.ndarray],
        float]]): Error function that needs to be optimized
        financial_instruments (FinancialInstruments): Financial instruments model object
        new_investment (float): Amount of new investment that needs to be distributed
        method (ScipyOptimizeMinimizeMethods, optional): Method to be used for
        minimizing the `error_function`. Defaults to ScipyOptimizeMinimizeMethods.SLSQP.

    Returns:
        np.ndarray: Returns array of split of `new_investment` into initial financial
        instruments to achieve target ratio as close as possible
    """
    num_investments: int = len(financial_instruments.instruments)
    error_function_callable: Callable[[np.ndarray], float] = error_function(financial_instruments, new_investment)
    constraints = [
        {"type": "ineq", "fun": _constraint_between_0_and_1},  # Inequality constraint: all elements between 0 and 1
        {"type": "eq", "fun": _constraint_sum_equal_to_1},  # Equality constraint: sum of elements equals 1
    ]
    bounds = tuple((0, 1) for _ in range(num_investments))  # Bounds for each variable: 0 to 1
    equitable_distribution = new_investment // num_investments
    x_0 = np.asarray([equitable_distribution] * (num_investments - 1) + [new_investment - (equitable_distribution * (num_investments - 1))])  # Initial guess

    results: scipy.optimize.OptimizeResult = scipy.optimize.minimize(error_function_callable, x_0, method=method.value, bounds=bounds, constraints=constraints)
    return results.x
