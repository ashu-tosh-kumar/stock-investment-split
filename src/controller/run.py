from src.controller.error import mean_squared_error
from src.controller.scipy_optimizer import scipy_optimize_minimize
from src.controller.utils.results import sanitize_too_small_values
from src.models.financial_instruments import FinancialInstruments
from src.models.optimizers import ScipyOptimizeMinimizeMethods


def run(financial_instruments: FinancialInstruments, new_investment: float) -> list[float]:
    """Main function to run the optimizer and return results

    Args:
        financial_instruments (FinancialInstruments): Model representing the financial
        models
        new_investment (float): Amount of new investment to be made

    Returns:
        list[float]: List of investment in each financial instrument to be made. It is
        in the same order as instruments in `financial_instruments.instruments` list
    """
    # Define financial instruments and their weightage
    print("Optimizing the MSE to compute optimal allocation. Sit tight!")
    financial_instruments_target_investment_array = scipy_optimize_minimize(
        mean_squared_error, financial_instruments, new_investment, method=ScipyOptimizeMinimizeMethods.SLSQP
    )

    # Sanitize too small values
    results_list = list(financial_instruments_target_investment_array)
    sanitize_too_small_values(results_list, inplace=True)

    return results_list
