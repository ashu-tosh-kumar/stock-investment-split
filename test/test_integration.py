from test.mocks import DUMMY_FINANCIAL_INSTRUMENTS

import pytest

from src.controller.run import run


@pytest.mark.integration
class TestStockInvestmentSplit:
    def test_stock_investment_split(self, snapshot):
        new_investment = 30

        actual_value = run(DUMMY_FINANCIAL_INSTRUMENTS, new_investment)

        assert snapshot == actual_value
