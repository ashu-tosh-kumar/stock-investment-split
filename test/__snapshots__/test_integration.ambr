# serializer version: 1
# name: TestStockInvestmentSplit.test_stock_investment_split
  list([
    10,
    10,
    10,
  ])
# ---
